README.md
# WhatsApp Search Bar for Flutter
  WhatsApp Search Bar its a Flutter plugin enable you to build a flexible search bar with buttons filter like WhatsApp Search bar

|<img src="https://raw.githubusercontent.com/mattar88/wp_search_bar/main/example/screenshots/1.png" width="245" hspace="4">
| <img src="https://raw.githubusercontent.com/mattar88/wp_search_bar/main/example/screenshots/2.png" width="245" hspace="4">
|<img src="https://raw.githubusercontent.com/mattar88/wp_search_bar/main/example/screenshots/3.png" width="245" hspace="4">|

## Usage
1.Install Package
With Dart:
```
$ dart pub add wp_search_bar
```
With Flutter:
```
$ flutter pub add wp_search_bar
```
This will add a line like this to you package's pubspec.yaml (and run an implicit `dart pub get`):
```
dependencies:
  wp_search_bar: ^0.1.4
```

2.Define data that you want to show in listview
```
  List<Map<String, String>> data = [
    {'name': "Mohammad", 'message': "message from mhammad"},
    {'name': "Kamel", 'message': "message from kamel"},
    {'name': "Nadine", 'message': "how are you ?"},
    {'name': "Joseph", 'message': "hello...!"},
    {'name': "71966691", 'message': "Hi"},
    {'name': "03405609", 'message': "hiiii...."},
    {'name': "Mohammad", 'message': "message from mhammad"},
    {'name': "Kamel", 'message': "message from kamel"},
    {'name': "Nadine", 'message': "how are you ?"},
    {'name': "Joseph", 'message': "hello...!"},
    {'name': "71966691", 'message': "Hi"},
    {'name': "03405609", 'message': "hiiii...."},
  ];
```

3.Define Button filters that should use in search bar.
```
var buttonFilters = {
          'name': {
            'name': 'name',
            'selected': false,
            'title': 'Name',
            'operation': 'CONTAINS',
            'icon': Icons.supervised_user_circle_rounded,
          },
          'message': {
            'name': 'message',
            'selected': false,
            'title': 'Message',
            'operation': 'CONTAINS',
            'icon': Icons.message,
          },
          'date': {
            'name': 'Date',
            'selected': false,
            'title': 'Date',
            'operation': 'CONTAINS',
            'icon': Icons.date_range,
          },
          'other': {
            'name': 'other',
            'selected': false,
            'title': 'Other',
            'operation': 'CONTAINS',
            'icon': Icons.more,
          },
        };
```
4.Build the Search bar
 
```
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: WPSearchBar(
        listOfFilters: buttonFilters,
        materialDesign: const {
          'title': {'text': 'WhatsApp'},
        },
        onSearch: (filter, value, operation) {
        },
        body: Container(
          decoration: const BoxDecoration(color: Color(0xff121b22)),
          child: ListView.builder(
              itemCount: filteredData.length,
              itemBuilder: (context, index) {
                var item = filteredData[index];
                return ListTile(
                  title: Text(
                    item['name'].toString(),
                    style: const TextStyle(
                        color: Colors.white, fontWeight: FontWeight.bold),
                  ),
                );
              }),
          ),
      ),
    );
  }
}

```

# To add Sidebar Menu

<table>
  <tr>
    <td align="center">Screenshot 1</td>
    <td align="center">Screenshot 2</td>
  </tr>
  <tr>
    <td><img alt="A Screenshot of WhatsApp Search Bar" src="https://raw.githubusercontent.com/mattar88/wp_search_bar/main/example/screenshots/sidebar_menu_closed.png" width="245"/></td>
    <td><img alt="Another Screenshot of WhatsApp Search Bar" src="https://raw.githubusercontent.com/mattar88/wp_search_bar/main/example/screenshots/sidebar_menu_open.png" width="245"/></td>
  </tr>
 </table>

### use a GlobalKey for scaffold
```
  final GlobalKey<ScaffoldState> _scaffoldKey = GlobalKey<ScaffoldState>();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        key: _scaffoldKey,
        endDrawer: SidebarMenu(),
        body: WPSearchBar(
          actions: [
            IconButton(
              icon: const Icon(Icons.menu),
              onPressed: () {
                _scaffoldKey.currentState?.openEndDrawer();
              },
            ),
          ],
          ....
        );
    }
```

CHANGELOG.md
## 0.0.1

* First release.

## 0.0.2

* Update readme.

## 0.0.3

* Update readme.
* Improve code.

## 0.0.4

* Update readme.

## 0.0.5

* Update readme.

## 0.0.6

* Update the usability of automaticallyImplyLeading option of AppBar

## 0.0.7

* Add background color for AppBar
* Improve Example
* Improve code and clean it

## 0.0.8

* Add event when Select/Close Filter and other options
* Improve Example

## 0.0.9

* Fix bugs for optional function
* Improve Documentation

## 0.1.1

* Fix bugs
* Clean Code

## 0.1.2

* pass progress indicator as given parameter widget

## 0.1.3

* Improve code

## 0.1.4

* Fix the problem right to left

test/wp_search_bar_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:wp_search_bar/wp_search_bar.dart';

void main() {
  test('adds one to input values', () {
    // final calculator = Calculator();
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
  });
}


example/lib/main.dart
import 'dart:developer';

import 'package:flutter/material.dart';
import 'package:wp_search_bar/wp_search_bar.dart';

void main() {
  runApp(const MyApp());
}

abstract class StatusEnum {
  ///Reached Whatsapp Servers
  static String rcs = "RCS";

  ///Reached Other Person
  static String rcp = "RCP";

  ///Read By other Person
  static String rdp = "RDP";
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  List<Map<String, dynamic>> data = [
    {'name': "Mohammad", 'message': "message from mhammad", 'status': 'RDP'},
    {'name': "Kamel", 'message': "message from kamel", 'status': 'RDP'},
    {'name': "Nadine", 'message': "how are you ?", 'status': 'RCP'},
    {'name': "Joseph", 'message': "hello...!", 'status': 'RCS'},
    {'name': "71966691", 'message': "Hi", 'status': 'RCP'},
    {'name': "03405609", 'message': "hiiii....", 'status': 'RCP'},
    {'name': "Mohammad", 'message': "message from mhammad", 'status': 'RCP'},
    {'name': "Kamel", 'message': "Hi mohammad", 'status': 'RCP'},
    {'name': "Nadine", 'message': "I am busy", 'status': 'RCS'},
    {'name': "Joseph", 'message': "hello there...!", 'status': 'RCP'},
    {'name': "71966691", 'message': "Hi there", 'status': 'RCS'},
    {'name': "03405609", 'message': "", 'status': 'RCS'},
  ];
  List<Map<String, dynamic>> filteredData = [];
  @override
  void initState() {
    filteredData = data;
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: WPSearchBar(
        appBarBackgroundColor: const Color(0xff1f2c34),
        listOfFilters: const {
          'name': {
            'name': 'name',
            'selected': false,
            'title': 'Name',
            'operation': 'CONTAINS',
            'icon': Icons.supervised_user_circle_rounded,
          },
          'message': {
            'name': 'message',
            'selected': false,
            'title': 'Message',
            'operation': 'CONTAINS',
            'icon': Icon(Icons.message),
          },
          'read_by_person': {
            'name': 'read_by_person',
            'selected': false,
            'title': 'Read by Other Person',
            'operation': 'CONTAINS',
            'icon': Icon(
              Icons.checklist_sharp,
              color: Colors.blue,
            ),
          },
          'date': {
            'name': 'Date',
            'selected': false,
            'title': 'Date',
            'operation': 'CONTAINS',
            'icon': Icon(Icons.date_range),
          },
        },
        materialDesign: {
          'title': const {'text': 'WhatsApp'},
          'loadingIndicator': Transform.scale(
            scale: 0.5,
            child: CircularProgressIndicator(
              strokeWidth: 3,
            ),
          ),
        },
        onSearch: (filterSelected, value, operation) {
          filteredData = data.where((Map<String, dynamic> element) {
            var dataKey = filterSelected;

            if (dataKey != null) {
              if (dataKey == 'read_by_person') {
                dataKey = 'name';
                if (element['status'] == null ||
                    element['status'] != (StatusEnum.rdp)) {
                  return false;
                }
              }
            }

            if (value == null || value == '') {
              return true;
            }

            print(
                'Element ${element[dataKey]}, filter: ${dataKey}, value: ${value}');
            if (dataKey != null &&
                element[dataKey] != null &&
                element[dataKey]!.toLowerCase().contains(value.toLowerCase())) {
              return true;
            } else {
              if (element['name']!
                      .toLowerCase()
                      .contains(value.toLowerCase()) ||
                  element['message']!
                      .toLowerCase()
                      .contains(value.toLowerCase())) {
                return true;
              }
            }

            return false;
          }).toList();
          setState(() {
            filteredData;
          });
        },
        body: Container(
          decoration: const BoxDecoration(color: Color(0xff121b22)),
          child: ListView.builder(
              itemCount: filteredData.length,
              itemBuilder: (context, index) {
                var item = filteredData[index];
                return ListTile(
                  leading: CircleAvatar(
                    child: Text(item['name']![0].toUpperCase()),
                    backgroundColor: const Color(0xff00a982),
                  ),
                  title: Text(
                    item['name'].toString(),
                    style: const TextStyle(
                        color: Colors.white, fontWeight: FontWeight.bold),
                  ),
                  subtitle: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(
                        item['message'].toString(),
                        style: const TextStyle(color: Color(0xff738088)),
                      ),
                      if (item['status'] == StatusEnum.rdp)
                        const Icon(
                          Icons.checklist_sharp,
                          color: Colors.blue,
                        ),
                      if (item['status'] == StatusEnum.rcs)
                        const Icon(
                          Icons.checklist_sharp,
                          color: Colors.grey,
                        ),
                      if (item['status'] == StatusEnum.rcp)
                        const Icon(Icons.check, color: Colors.grey, size: 16),
                    ],
                  ),
                  trailing: const Text(
                    '1:30 am',
                    style: TextStyle(color: Color(0xff738088)),
                  ),
                );
              }),
        ),
      ),
    );
  }
}
